File: scratch/extracted/graph-dfs/optimize.cpp
// Definition for a graph node (if needed).
// struct GraphNode {
//     int val;
//     vector<GraphNode*> neighbors;
//     GraphNode(int x) : val(x) {}
// };

class Solution {
private:
    /**
     * Recursive helper function to perform Depth-First Search starting from a given node.
     */
    void explore(vector<vector<int>>& graph, int node,
                 unordered_set<int>& visited, vector<int>& result) {

        // Mark current node as visited
        visited.insert(node);

        // Add node to result
        result.push_back(node);

        // Iterate through neighbors
        for (int neighbor : graph[node]) {
            // If neighbor not visited, explore recursively
            if (visited.find(neighbor) == visited.end()) {
                explore(graph, neighbor, visited, result);
            }
        }
    }

public:
    /**
     * Performs Depth-First Search starting from a given node
     */
    vector<int> dfs(vector<vector<int>>& graph, int start) {

        unordered_set<int> visited;
        vector<int> result;

        explore(graph, start, visited, result);

        return result;
    }
};

File: scratch/extracted/graph-dfs/optimize.ts
function dfs(graph: number[][], start: number): number[] {
  // visited set to keep track of visited nodes.
  const visited = new Set<number>();
  // result array stores the order of visited nodes.
  const result: number[] = [];

  // Recursive helper function to explore the graph.
  function explore(node: number) {
    // Mark the current node as visited.
    visited.add(node);
    // Add the current node to the result.
    result.push(node);

    // Iterate through the neighbors of the current node.
    for (const neighbor of graph[node]) {
      // If the neighbor has not been visited,
      if (!visited.has(neighbor)) {
        // Recursively explore the neighbor.
        explore(neighbor);
      }
    }
  }

  // Start the DFS traversal from the given start node.
  explore(start);
  // Return the result array.
  return result;
}